### PR TITLE
Framework as package

### LINKED ISSUE #42: Consistent Module Imports
--------
Issue Description
--------
##### What did you expect to see happen?
Consistent, modular imports throughout RAVEN

##### What did you see instead?
Fiddling with PYTHONPATH and some inconsistency in when some modules can import others safely, as well as occasional imports in the body of python modules

##### Do you have a suggested fix for the development team?
We should probably continue towards making RAVEN more library-like, so that our imports all happen in a consistent way.

----------------
For Devs: Issue Review
----------------
- [x] 1. Is it tagged with a type: defect or improvement? __tagged__
- [x] 2. Is it tagged with a priority: critical, normal or minor? __tagged__
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements? __n/a__
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users. __n/a__
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.) __provided__

-------
For Devs: Issue Closure
-------
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by joshua-cogliati-inl**
--------
Pull Request Description
--------
##### What issue does this change request address?
#1764
Closes #42

##### What are the significant changes in functionality due to this change request?
This changes framework to a package.  This allows creating a pip package that just has a framework directory in the root directory, instead of multiple directories and files.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by joshua-cogliati-inl** [comment]
Suggested email:

In order to make a raven pip package, RAVEN's code needs to be inside a single package.  This requires changing all the imports, since now they are inside ravenframework, instead of just at the top level.  In addition, the directory framework has been renamed to ravenframework.

This will require changes in three things.
1. Any imports of things that used to be in framework, now need to have ravenframework explicitly mentioned.
2. Direct use of Driver.py can no longer be done, since the imports will not be setup correctly.  Use raven_framework.py instead.
3. Pickle files have the package information embedded inside, so they will need to be regenerated.

The following are example modifications for framework as a package in the source code.



Relative (use . if in ravenframework/, .. if in ravenframework/someDir/, ... if in ravenframework/someDir/anotherDir/ and so forth.



Example for something that is in a directory inside of ravenframework:



Original:

import Files

New:

from .. import Files



Original:

from utils import mathUtils

New:

from ..utils import mathUtils



Original:

import utils.TreeStructure

New:

from ..utils import TreeStructure

(and modify all uses of utils.TreeStructure to TreeStructure)



Original:

import utils.TreeStructure as TS

New:

from ..utils import TreeStructure as TS





Ravenframework imports:



Original:

import Files

New:

from ravenframework import Files



Original:

from utils import mathUtils

New:

from ravenframework.utils import mathUtils



Original:

import utils.TreeStructure

New:

from ravenframework.utils import TreeStructure

(and modify all uses of utils.TreeStructure to TreeStructure)



Original:

import utils.TreeStructure as TS

New:

from ravenframework.utils import TreeStructure as TS







Importlib (Example from ravenframework/Models/Code.py )

Original:

importlib.import_module("CodeInterfaces")

New:

importlib.import_module("..CodeInterfaces", "ravenframework.Models")

(Note the second parameter package anchor for finding relative packages.)





Driver.py can no longer be used from the python executable.



Original:

python framework/Driver.py "${ARGS[@]}"

New:

python raven_framework.py "${ARGS[@]}"

or

raven_framework "${ARGS[@]}"

**Comment by joshua-cogliati-inl** [comment]
Before this is merged, we should send an email warning people about the upcoming change.

**Comment by PaulTalbot-INL** [comment]
I see there's a TEAL update in there; that's intentional? Just checking for completeness.

File: ravenframework/InputTemplates/TemplateBaseClass.py
Comment (PaulTalbot-INL) [inline ravenframework/InputTemplates/TemplateBaseClass.py:30]: We can probably remove these lines; would they be instructive to leave in?

Comment (joshua-cogliati-inl) [reply ravenframework/InputTemplates/TemplateBaseClass.py:30]: I'll remove them.

**Review by PaulTalbot-INL — DISMISSED** [review]
I recommend we remove the TEAL update and the two commented lines; otherwise the changes pass code check.

**Comment by PaulTalbot-INL** [comment]
I think this PR will need @wangcj05 control board approval.

**Comment by PaulTalbot-INL** [comment]
I don't think these changes will impact users beyond the email Joshua prepared. However, I think this significantly impacts our approaches as developers, so I wonder if a Monday or Wednesday meeting should cover what the impact of these changes are on:

1. Installing RAVEN (and plugins)

   1. Installation via pip
    2. Installing and running tests (separate?) 

2. Pathing within RAVEN (and plugins)
3. ? Anything else I missed

**Comment by joshua-cogliati-inl** [comment]
> Installing RAVEN (and plugins)
> 
>     1. Installation via pip
> 
>     2. Installing and running tests (separate?)

At the moment, the install is still the same.  I am continuing to work on making a pip package work at: https://github.com/joshua-cogliati-inl/raven/tree/package_testing

File: ravenframework/Metrics/metrics/DSS.py
Comment (wangcj05) [inline ravenframework/Metrics/metrics/DSS.py:26]: @joshua-cogliati-inl which is a better way to import the module, "ravenframework.utils" or "...utils"? I think I prefer to using "ravenframework.utils"

Comment (joshua-cogliati-inl) [reply ravenframework/Metrics/metrics/DSS.py:26]: Hm, well, one reason I used ...utils was so I didn't have to change it when framework got renamed.  Right now ravenframework.whatever is a good warning that that section of code has something strange happening to the python path (like in CodeInterfaces or in CustomModes). For this pull request I recommend keeping the relative paths, but I have no strong opinions on this for future additions.

**Comment by joshua-cogliati-inl** [comment]
> 
> I agree with @PaulTalbot-INL, I think next Monday is good for the discussion. @joshua-cogliati-inl Can you prepare some slides for next Monday discussion?

Okay, I can make some slides for next Monday's discussion.

P.S. I will look thru the manual for imports and use of Driver.

**Comment by joshua-cogliati-inl** [comment]
I updated the user manual fixing places that used import and Driver.py.

**Review by wangcj05 — APPROVED** [review]
PR Looks good.

**Comment by wangcj05** [comment]
PR is good and checklist is satisfied.

**Comment by wangcj05** [comment]
@joshua-cogliati-inl Could you send out an email to the users and plugin developers (LOGOS, TEAL, HERON, FARM, etc.)?

**Comment by wangcj05** [comment]
Wonderful! Thanks a lot.

Congjian
________________________________
From: Haoyu Wang ***@***.***>
Sent: Thursday, May 12, 2022 12:12 PM
To: idaholab/raven ***@***.***>
Cc: Congjian Wang ***@***.***>; State change ***@***.***>
Subject: [EXTERNAL] Re: [idaholab/raven] Framework as package (PR #1784)


Hello Congjian,
Thank you for letting me know.
I can fix it. Paul gave me the instructions on how to update the submodule
versions a moment ago. I will get it done no later than next Monday.
Thanks,
Haoyu

**Comment by wanghy-anl** [comment]
Hello Congjian,
I opened a PR #1826 called "Farm submodule update". Please have a check and
let me know if more changes are needed.
Thanks,
Haoyu

**Comment by wangcj05** [comment]
Hi Haoyu,

Thanks. Once FARM pass the regression tests, I will merge it.

Best,

Congjian
________________________________
From: Haoyu Wang ***@***.***>
Sent: Thursday, May 12, 2022 9:28 PM
To: idaholab/raven ***@***.***>
Cc: Congjian Wang ***@***.***>; State change ***@***.***>
Subject: [EXTERNAL] Re: [idaholab/raven] Framework as package (PR #1784)


Hello Congjian,
I opened a PR #1826 called "Farm submodule update". Please have a check and
let me know if more changes are needed.
Thanks,
Haoyu

**Pull request merged**
